### PR TITLE
koordlet:move audit log to resource executor after resource update

### LINKED ISSUE #1018: [proposal]move audit log to resource executor
**What is your proposal**:
It is expected to solve the problem of duplicate printing logs of audit logs mentioned  by moving it to the executor module, since the module supports the cacheable method of r/w.

**Why is this needed**:

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by Re-Grh**
### Ⅰ. Describe what this PR does
to solve the problem of duplicate audit logs by moving it to the executor module after resource update

### Ⅱ. Does this pull request fix one issue?
fixes #1018
The audit log redundancy issue has been addressed by leveraging the executor's built-in cache, which has changed the recording frequency from 10 seconds to 60 seconds. Meanwhile, the adjustbecpuset policy execution shows frequent CPU rebinding , both klog and audit log will print.

截屏2023-03-10 10 15 44

and klog shows here

截屏2023-03-10 10 44 33

and i will start a new issue since it looks like we need to modify the function applyCPUSetWithNonePolicy(beCPUSet, oldCPUSet) to improve the stability of CPUs for the BE Group's calculation policy.

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/resmanager/cpu_suppress.go [hunk 1]
```diff
@@ -434,7 +437,12 @@ func (r *CPUSuppress) adjustByCfsQuota(cpuQuantity *resource.Quantity, node *cor
 		newBeQuota = *currentBeQuota + int64(beMaxIncreaseCPUQuota)
 	}
 
-	updater, err := resourceexecutor.DefaultCgroupUpdaterFactory.New(system.CPUCFSQuotaName, beCgroupPath, strconv.FormatInt(newBeQuota, 10))
+	auditdata := &audit.Event{
+		Type:    "node",
```
Comment (saintube) [inline pkg/koordlet/resmanager/cpu_suppress.go:441]: should be defined in the resource executor package (e.g. `Type: resourceexecutor.NodeEvent`)

File: pkg/koordlet/resmanager/cpu_suppress.go [hunk 2]
```diff
@@ -356,7 +357,9 @@ func (r *CPUSuppress) adjustByCPUSet(cpusetQuantity *resource.Quantity, nodeCPUI
 		klog.Warningf("suppressBECPU failed to apply be cpu suppress policy, err: %s", err)
 		return
 	}
-	_ = audit.V(1).Node().Reason(resourceexecutor.AdjustBEByNodeCPUUsage).Message("update BE group to cpuset: %v", beCPUSet).Do()
+	if !reflect.DeepEqual(beCPUSet, oldCPUSet) {
+		_ = audit.V(1).Node().Reason(resourceexecutor.AdjustBEByNodeCPUUsage).Message("update BE group to cpuset: %v", beCPUSet).Do()
```
Comment (saintube) [inline pkg/koordlet/resmanager/cpu_suppress.go:361]: should move into the resource updater of the `applyBESuppressCPUSet`

File: pkg/koordlet/resourceexecutor/updater.go [hunk 3]
```diff
@@ -152,6 +156,36 @@ func (u *CgroupResourceUpdater) SetUpdateFunc(updateFunc UpdateFunc, mergeUpdate
 	u.mergeUpdateFunc = mergeUpdateFunc
 }
 
+// CgroupFileWriteIfDifferent writes the cgroup file if current value is different from the given value.
+func (u *CgroupResourceUpdater) cgroupFileWriteIfDifferent() (bool, error) {
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/updater.go:160]: The naming `(*CgroupResourceUpdater).cgroupFileWriteIfDifferent` is wordy. And should all CgroupResourceUpdater implement this method? How about moving the invoke of auditor into the update functions?

File: pkg/koordlet/resourceexecutor/updater.go [hunk 4]
```diff
@@ -444,3 +521,18 @@ func MergeConditionIfCPUSetIsLooser(oldValue, newValue string) (string, bool, er
 	merged := v.Union(old)
 	return merged.String(), true, nil
 }
+
+func ParseEventForAuditLog(event *audit.Event, verbose int) {
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/updater.go:525]: This function does not only parse the event but also does the logging. We can call it `AuditLogForEvent()`.

File: pkg/koordlet/resourceexecutor/updater.go [hunk 5]
```diff
@@ -227,6 +261,7 @@ type NewResourceUpdaterFunc func(resourceType sysutil.ResourceType, parentDir st
 type ResourceUpdaterFactory interface {
 	Register(g NewResourceUpdaterFunc, resourceTypes ...sysutil.ResourceType)
 	New(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error)
+	NewWithAudit(resourceType sysutil.ResourceType, parentDir string, value string, auditdata *audit.Event, verbose int) (ResourceUpdater, error)
```
Comment (saintube) [inline pkg/koordlet/resourceexecutor/updater.go:264]: is there any case that we new an updater without the audit?

File: pkg/koordlet/resmanager/cpu_suppress.go [hunk 6]
```diff
@@ -81,9 +81,10 @@ func (r *CPUSuppress) RunInit(stopCh <-chan struct{}) error {
 // writeBECgroupsCPUSet writes the be cgroups cpuset by order
 func (r *CPUSuppress) writeBECgroupsCPUSet(paths []string, cpusetStr string, isReversed bool) {
 	var updaters []resourceexecutor.ResourceUpdater
+	eventHelper := audit.V(3).Reason(resourceexecutor.AdjustBEByNodeCPUUsage).Message("update BE group to cpuset: %v", cpusetStr)
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_suppress.go:84]: what is the difference between V(3) and V(1)?
we need a standard for this

File: pkg/koordlet/resourceexecutor/updater.go [hunk 7]
```diff
@@ -325,10 +358,29 @@ func (f *CgroupUpdaterFactoryImpl) New(resourceType sysutil.ResourceType, parent
...
+	c := u.(*CgroupResourceUpdater)
+	c.eventHelper = e
+	return c, nil
+}
+
 func CommonCgroupUpdateFunc(resource ResourceUpdater) error {
 	c := resource.(*CgroupResourceUpdater)
-	_ = audit.V(5).Reason(ReasonUpdateCgroups).Message("update %v to %v", resource.Path(), resource.Value()).Do()
-	return sysutil.CgroupFileWriteIfDifferent(c.parentDir, c.file, c.value)
+	dirty, err := cgroupFileWriteIfDifferent(c.parentDir, c.file, c.value)
```
Comment (zwzhang0107) [inline pkg/koordlet/resourceexecutor/updater.go:374]: dirty->updated

File: pkg/koordlet/resourceexecutor/updater.go [hunk 8]
```diff
@@ -227,6 +256,7 @@ type NewResourceUpdaterFunc func(resourceType sysutil.ResourceType, parentDir st
 type ResourceUpdaterFactory interface {
 	Register(g NewResourceUpdaterFunc, resourceTypes ...sysutil.ResourceType)
 	New(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error)
+	NewWithAudit(resourceType sysutil.ResourceType, parentDir string, value string, e *audit.EventHelper) (ResourceUpdater, error)
```
Comment (zwzhang0107) [inline pkg/koordlet/resourceexecutor/updater.go:259]: there is still some V5 logs in other pkg, which need to be modified.

File: pkg/koordlet/resmanager/cpu_suppress.go (hunk 6 above)
Comment (Re-Grh) [reply pkg/koordlet/resmanager/cpu_suppress.go:84]: temporarily use level 3 as the agreed-upon level, and define different audit levels based on needs later

**Comment by zwzhang0107** [comment]
problem about applyCPUSetWithNonePolicy?

File: pkg/koordlet/resourceexecutor/resctrl_updater.go [hunk 9]
```diff
@@ -151,7 +151,7 @@ func UpdateResctrlTasksFunc(resource ResourceUpdater) error {
 	// 122
 	// 123
 	// 124
-	_ = audit.V(5).Reason(ReasonUpdateResctrl).Message("update %v to %v", resource.Key(), resource.Value()).Do()
+	_ = audit.V(3).Reason(ReasonUpdateResctrl).Message("update %v to %v", resource.Key(), resource.Value()).Do()
```
Comment (zwzhang0107) [inline pkg/koordlet/resourceexecutor/resctrl_updater.go:154]: V5

File: pkg/koordlet/resmanager/cpu_suppress.go [hunk 10]
```diff
@@ -351,13 +353,20 @@ func (r *CPUSuppress) adjustByCPUSet(cpusetQuantity *resource.Quantity, nodeCPUI
 	// the new be suppress always need to apply since:
 	// - for a reduce of BE cpuset, we should make effort to protecting LS no matter how huge the decrease is;
 	// - for a enlargement of BE cpuset, it is welcome and costless for BE processes.
-	err = r.applyBESuppressCPUSet(beCPUSet, oldCPUSet)
-	if err != nil {
-		klog.Warningf("suppressBECPU failed to apply be cpu suppress policy, err: %s", err)
-		return
+	if !compareIntSlice(beCPUSet, oldCPUSet) {
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/cpu_suppress.go:356]: remove this

File: pkg/koordlet/resourceexecutor/updater.go [hunk 11]
```diff
@@ -315,25 +320,39 @@ func (f *CgroupUpdaterFactoryImpl) Register(g NewResourceUpdaterFunc, resourceTy
...
-func (f *CgroupUpdaterFactoryImpl) New(resourceType sysutil.ResourceType, parentDir string, value string) (ResourceUpdater, error) {
+func (f *CgroupUpdaterFactoryImpl) New(resourceType sysutil.ResourceType, parentDir string, value string, e *audit.EventHelper) (ResourceUpdater, error) {
 	f.lock.RLock()
 	defer f.lock.RUnlock()
 	g, ok := f.registry[resourceType]
 	if !ok {
 		return nil, fmt.Errorf("resource type %s not registered", resourceType)
 	}
-	return g(resourceType, parentDir, value)
+	u, err := g(resourceType, parentDir, value)
```
Comment (zwzhang0107) [inline pkg/koordlet/resourceexecutor/updater.go:330]: pass EventHelper to g()

File: pkg/koordlet/resourceexecutor/updater.go [hunk 12]
```diff
@@ -315,25 +320,39 @@ func (f *CgroupUpdaterFactoryImpl) Register(g NewResourceUpdaterFunc, resourceTy
...
+	}
+	if updated && c.eventHelper.Event.Reason != "" {
+		_ = c.eventHelper.Do()
+	} else if updated {
+		_ = audit.V(3).Reason(ReasonUpdateCgroups).Message("update %v to %v", resource.Path(), resource.Value()).Do()
+	}
+	return nil
 }
 
 func CommonDefaultUpdateFunc(resource ResourceUpdater) error {
```
Comment (zwzhang0107) [inline pkg/koordlet/resourceexecutor/updater.go:353]: NewCommonDefaultUpdater

File: pkg/koordlet/resmanager/system_config.go [hunk 13]
```diff
@@ -81,7 +81,7 @@ func caculateMemoryConfig(strategy *slov1alpha1.SystemStrategy, nodeMemory int64
 		if sysutil.ValidateResourceValue(&minFreeKbytes, "", sysutil.MinFreeKbytes) {
 			valueStr := strconv.FormatInt(minFreeKbytes, 10)
 			file := sysutil.MinFreeKbytes.Path("")
-			resource, err := resourceexecutor.NewCommonDefaultUpdater(file, file, valueStr)
+			resource, err := resourceexecutor.NewCommonDefaultUpdater(file, file, valueStr, nil)
```
Comment (zwzhang0107) [inline pkg/koordlet/resmanager/system_config.go:84]: should pass eventHelper, not nil

**Comment by zwzhang0107** [comment]
check the final result

**Pull request merged**